Appbar icons change appearance to match activeroute

ftanekha opened the pull request:
Closes #3

ftanekha committed:
Debugging:
- TypeError: SearchContext.Provider component had a value prop of any
- fix: provide the value prop with a specific type of Partial<{ status, setStatus, query, setQuery }>, making all of its properties are now optional

ftanekha committed:
- provide a not-found error boundary for word search with delayed automatic redirect to homepage

ftanekha committed:
- undo package.json changes
- remove not-found error boundary

ftanekha committed:
add title to search button

ftanekha committed:
- AppBar nav items change apperance to indicate the current/active route
- refactor nav items classLists for DRY purposes

ftanekha committed:
remane appBarItemClassList to appBarItemClassList, to match component name

asideofcode-dev commented on src/component/appbar.component.tsx:
```diff
@@ -15,23 +16,24 @@ const inter = Inter({
           <Link href="/">
             <SvgIcon
-              className="h-6 w-6 cursor-pointer hover:text-blue-500 hover:scale-110 transition-transform duration-200"
+              className={`${appBarItemClassList} ${pathname === '/' && 'active'}`}
```
Nice, this does exactly as the issue requested! 

I would like to ask for 2 improvements:

1. I think this is great opportunity to use the `classnames`  library, see https://github.com/asideofcode/duramazwi/blob/main/src/component/atom/loader.component.tsx#L6 for an example. At the moment when the the condition short-circuits it adds the string `"false"` to the `className`, which isn't ideal.
2. Right now `"active"` gets added to the `className` but there's no visible component. Let's use something like `text- blue-600` which is visually distinctive!

asideofcode-dev commented on src/component/search-bar.component.tsx:
```diff
@@ -60,7 +60,7 @@ export default function SearchBar({}:
           <Loading className="h-6 w-6" />
         ) : (
-          <button className="h-6 w-6 theme-text-sub1" onClick={search}>
+          <button className="h-6 w-6 theme-text-sub1" title="search" onClick={search}>
```
Nice! I think it might valuable to update this and the other titles to use title-case ?

ftanekha replied:
yea, I hadn't seen this comment. I actually tried to call you last night to find out what case you were refering to in your other comment

asideofcode-dev reviewed (changes requested):
That was quick, thanks! It does as the issue requested but I've left some comments with suggestion for improvements.

Separately, there are too many commits in the commit history of this PR branch at the moment, it might useful to squish them. This SO post (https://stackoverflow.com/a/58082749) points to how you might do it. I could always squash the commits on merge but might be good to get some practice on reorganising your commit history!

ftanekha committed:
- give search button more specific title
- provide AppBar items with trimmed classList

ftanekha committed:
- leverage React **classnames** _module_ to conditionally replace 'active' classname with corresponding _TailwindCss_ classname 'text-blue-600'
- change SvgIcon className prop's type from string to boolean
- apply SvgIcon style.display as inline TailwindCss class 'inline-block' instead of through style attribute i.e. style={{ display: "inline-block" }}
- minor refactoring/ clean-up

ftanekha committed:
remove unused file

ftanekha committed:
minor refactoring

asideofcode-dev commented on src/component/icons/svg-icon.tsx:
```diff
@@ -59,8 +60,14 @@ const SvgIcon = ({
+      {
+        getSVGIcon(
+          icon, size, 
+          cx('inline-block h-6 w-6 cursor-pointer hover:text-blue-500 hover:scale-110 transition-transform duration-200', {'text-blue-600': className}), 
```
I think I see where you're going with this, and I think the instinct is good. The same list of classes is being used across the instances of  `SvgIcon` and you've consolidated them. The issue is that now `SvgIcon` is fundamentally tied to that list of classes. I think a more flexible approach with similar benefits would be to maintain `SvgIcon` as is and perhaps define a more specialised SvgIcon instance, that can be used in all those places.

```ts
import cx from 'classnames';

type SvgIconButtonProps = {
  icon: Icons;
  className?: string;
  onClick?: (e: any) => void;
  size?: SvgIconSize;
  variant?: any;
  title?: string;
};

type StyledSvgIconProps = SvgIconButtonProps & { active?: boolean };

function StyledSvgIcon({ className, active, ...restOfProps }: StyledSvgIconProps) {
  return (
    <SvgIcon
      {...restOfProps}
      className={cx(
        'inline-block h-6 w-6 cursor-pointer hover:text-blue-500 hover:scale-110 transition-transform duration-200',
        { 'text-blue-600': active },
        className
      )}
    />
  );
}

```

asideofcode-dev commented on src/component/search-bar.component.tsx:
```diff
@@ -57,16 +57,17 @@ export default function SearchBar({}:
+          <Loading className="h-6 w-6"/>
+            : 
+          (
+            <button className="h-6 w-6 theme-text-sub1" title="search word" onClick={search}>
```
In my earlier comment I mentioned using title-case for this Button and the other ones that you'd changed in a previous PR. So something like `"Search word"` vs `"search word"`.

asideofcode-dev reviewed (changes requested):
Loving the progress! Good to see `classnames` showing up. 2 things left to address before we proceed.

ftanekha committed:
- address AppBar svgicons title attribute case issue i.e capitalise values
- implement styled SvgIcon with extra prop (ative), reflecting route changes

ftanekha committed:
- provide a not-found error boundary with delayed automatic redirect to homepage

asideofcode-dev commented on src/app/not-found.tsx:
```diff
@@ -0,0 +1,14 @@
+'use client'
+import { redirect } from 'next/navigation'
+
+function ErrorBoundary() {
```
Why is this file here, assuming it somehow slipped through the net? Otherwise remove

ftanekha replied:
I used the not-found boundary because the 404 page you had on there had a text color that difficult to read in dark mode. And i couldn't locate the code for it in the files to make the necessary changes. I thought it was a good touch because the message is clear, but I've removed the automatic redirect

asideofcode-dev commented on src/component/appbar.component.tsx:
```diff
@@ -4,49 +4,51 @@ import { Inter } from "next/font/google";
   display: "swap",
 });
 
+const defaultSvgIconClassList = 'inline-block h-6 w-6 cursor-pointer hover:text-blue-500 hover:scale-110 transition-transform duration-200';
```
This class should just be set on the styled component. This default class name + the the `active` prop is the reason why we wanted to define the styled component. Please look at my previous comment https://github.com/asideofcode/duramazwi/pull/9#discussion_r1993133944.

asideofcode-dev commented on src/component/icons/styled-appbar-svg-icon.tsx:
```diff
@@ -0,0 +1,18 @@
+        }  
+    />  
+  );  
+}
```
Please add a new line at the end of any files you add (all of them, the pre-existing ones have the newline already), see https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline for why.

asideofcode-dev reviewed (changes requested):
Few more changes

ftanekha committed:
- set a default classname for styled SVGicons
- remove the automatic redirect in the not-found boundary for the root directory

asideofcode-dev commented on src/app/not-found.tsx:
```diff
@@ -0,0 +1,11 @@
+            This page does not exist!
+        </h1>
+    )
+}
```
Add new line at EOF

asideofcode-dev reviewed (approved):
LGTM!
